app/processing/__init__.py

```diff
@@ -1,3 +1,77 @@
 """
 Provides the core logic of the backend.
 """
+
+from app.model.CompileRequest import (
+    CompileRequest,
+    ImplementationNode,
+)
+from app.model.CompileRequest import (
+    Node as FrontendNode,
+)
+from app.openqasm3.printer import leqo_dumps
+from app.processing.graph import (
+    IOConnection,
+    ProgramGraph,
+    ProgramNode,
+)
+from app.processing.merge import merge_nodes
+from app.processing.optimize import optimize
+from app.processing.post import postprocess
+from app.processing.pre import preprocess
+
+
+class _Processor:
+    request: CompileRequest
+    graph: ProgramGraph
+    lookup: dict[str, tuple[ProgramNode, FrontendNode]]
+
+    def __init__(self, request: CompileRequest):
+        self.request = request
+        self.graph = ProgramGraph()
+        self.lookup = {}
+
+    def process(self) -> str:
+        for frontend_node in self.request.nodes:
+            program_node = ProgramNode(frontend_node.id)
+            self.lookup[frontend_node.id] = (program_node, frontend_node)
+
+            if not isinstance(frontend_node, ImplementationNode):
+                raise TypeError(
+                    f"Type of node {frontend_node.id} must be ImplementationNode"
+                )
```

Comment:
I would move the error handling. For me it is cleaner to first error handle and then do stuff.

```suggestion
            if not isinstance(frontend_node, ImplementationNode):
                raise TypeError(
                    f"Type of node {frontend_node.id} must be ImplementationNode"
              )

            program_node = ProgramNode(frontend_node.id)
            self.lookup[frontend_node.id] = (program_node, frontend_node)
```